### PR TITLE
Inasafe #1811 - New implementation for analysis extents definition

### PR DESCRIPTION
**Pull request by timlinux**
Please see the screenshots and video on #1811 for details. I am still polishing a few things (help, removal of old options), but would appreciate any early feedback from those interested.

### CODE REVIEW

**Comment by akbargumbira** [comment]
test this please, jenkins

**Comment by timlinux** [comment]
test this please, jenkins

**Comment by akbargumbira** [comment]
Hi @timlinux, it seems good. But it seems off when I use volcano point layer.

Another thing with methods in qgis_utilities.py, I think we can have one method: 
```display_message_bar(iface, message_type, title, message, more_details, button_text, duration)```
with
```message_type``` in ``` [QgsMessageBar.INFO, QgsMessageBar.SUCCESS, QgsMessageBar.WARNING, QgsMessageBar.CRITICAL]```

to cover all the methods for showing message bar..

File: safe/gui/ui/dock_base.ui [hunk 1]
```diff
@@ -57,13 +60,16 @@
         <number>0</number>
        </property>
        <property name="verticalSpacing">
-        <number>1</number>
+        <number>10</number>
```
Comment (akbargumbira) [inline safe/gui/ui/dock_base.ui]: not too much vertical spacing?

File: safe/gui/widgets/dock.py [hunk 2]
```diff
@@ -1900,12 +1901,43 @@ def _has_active_layer(self):
 
         .. versionadded:: 3.1
 
-        :returns: True if there is a layer highlighted in the legend.
+        :returns: True if there is a layer hightlighted in the legend.
```
Comment (akbargumbira) [inline safe/gui/widgets/dock.py]: http://www.merriam-webster.com/dictionary/highlight

**Comment by Charlotte-Morgan** [comment]
@timlinux - i tested with flood raster on population in Jakarta and got a py error + an error in the dock

Error message
```
An error has occured while executing Python code:


Traceback (most recent call last):
  File "C:/Users/AIFDR_ADMIN/.qgis2/python/plugins\inasafe\safe\gui\widgets\dock.py", line 741, in on_cboHazard_currentIndexChanged
    self.get_functions()
  File "C:/Users/AIFDR_ADMIN/.qgis2/python/plugins\inasafe\safe\gui\widgets\dock.py", line 1069, in get_functions
    raise e
IOError: [Errno 9] Bad file descriptor


Python version:
2.7.5 (default, May 15 2013, 22:44:16) [MSC v.1500 64 bit (AMD64)]


QGIS version:
2.8.1-Wien Wien, exported

Python path: ['C:/PROGRA~1/QGISWI~1/apps/qgis/./python/plugins\\processing', 'C:/PROGRA~1/QGISWI~1/apps/qgis/./python', u'C:/Users/AIFDR_ADMIN/.qgis2/python', u'C:/Users/AIFDR_ADMIN/.qgis2/python/plugins', 'C:/PROGRA~1/QGISWI~1/apps/qgis/./python/plugins', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\matplotlib-1.3.1-py2.7-win-amd64.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\nose-1.3.3-py2.7.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\tornado-4.0.1-py2.7-win-amd64.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\backports.ssl_match_hostname-3.4.0.2-py2.7.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\certifi-14.05.14-py2.7.egg', 'C:\\Program Files (x86)\\ArcGIS\\Desktop10.2\\bin', 'C:\\Program Files (x86)\\ArcGIS\\Desktop10.2\\arcpy', 'C:\\Program Files (x86)\\ArcGIS\\Desktop10.2\\ArcToolbox', 'C:\\PROGRA~1\\QGISWI~1\\bin\\python27.zip', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\DLLs', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\plat-win', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\lib-tk', 'C:\\PROGRA~1\\QGISWI~1\\bin', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\PIL', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\jinja2-2.7.2-py2.7.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\markupsafe-0.23-py2.7-win-amd64.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\pytz-2012j-py2.7.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\win32', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\win32\\lib', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\Pythonwin', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\Shapely-1.2.18-py2.7-win-amd64.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\six-1.3.0-py2.7.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\wx-2.8-msw-unicode', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\xlrd-0.9.2-py2.7.egg', 'C:\\PROGRA~1\\QGISWI~1\\apps\\Python27\\lib\\site-packages\\xlwt-0.7.5-py2.7.egg', u'C:/Users/AIFDR_ADMIN/.qgis2//python', 'C:/Users/AIFDR_ADMIN/.qgis2/python/plugins\\inasafe', 'C:\\Users\\AIFDR_ADMIN\\.qgis2\\python\\plugins\\inasafe\\safe_extras\\parameters', 'C:\\Users\\AIFDR_ADMIN\\.qgis2\\python\\plugins', 'C:\\Users\\AIFDR_ADMIN\\.qgis2\\python\\plugins\\inasafe\\safe_extras', 'C:\\PROGRA~1\\QGISWI~1\\apps\\qgis\\python\\plugins\\fTools\\tools', 'C:/WorkSpace/InaSAFE Testing']
```
Message in dock
image

**Comment by Charlotte-Morgan** [comment]
another option for this help is to make sure the hazard data and exposure data are turned on
image

**Comment by Charlotte-Morgan** [comment]
in this dialogue box - does the user expect to get confirmation and warning messages when extents are valid / not valid if they choose one of the top three options? If not - these options should be greyed out and only operational when the bounding box is selected 
image

**Comment by Charlotte-Morgan** [comment]
i like the warning / validation bar and the help dialogue :)

File: safe/gui/ui/dock_base.ui (hunk 1 above)
Comment (timlinux) [inline safe/gui/ui/dock_base.ui]: How does it look for you? For mac it fixes the squashed up look in the dock.

**Comment by timlinux** [comment]
test this please, jenkins

**Comment by timlinux** [comment]
Just a belated response here @Charlotte-Morgan 

> in this dialogue box - does the user expect to get confirmation and warning messages when extents are valid / not valid if they choose one of the top three options? If not - these options should be greyed out and only operational when the bounding box is selected 

The extent selector does not know *apriori* what layers will be used - an appropriate message should be show at run time via the message bar if the extents are not valid.

**Comment by timlinux** [comment]
Just an update to @akbargumbira 's comment - I tested with volcano point and it seems to work ok for me:

qgis 2 8 3-wien - test_003


qgis 2 8 3-wien - test_005

**Comment by timlinux** [comment]
In reply to @Charlotte-Morgan : 

> @timlinux - i tested with flood raster on population in Jakarta and got a py error + an error in the dock

I believe this is working fine.

**Comment by timlinux** [comment]
@ismailsunni please review so I can merge tomorrow. Thanks!

File: safe/test/data/control/files/test-full-run-results-qgis.txt [hunk 3]
```diff
@@ -15,16 +15,16 @@ Exposure details
 Population - sourced from Not specified
 Detailed gender report (affected people)
 Kab_name Total Female count (affected) Weekly hygiene packs Additional weekly rice kg for pregnant and lactating women
-JAKARTA PúSAT 93 46 37 4
+SAT 93 46 37 4
```
Comment (ismailsunni) [inline safe/test/data/control/files/test-full-run-results-qgis.txt]: Why do you modify it? I think the unicode char is needed for the test.

File: safe/utilities/analysis.py [hunk 4]
```diff
@@ -391,9 +378,31 @@ def get_clip_parameters(self):
...
+            'HazardExposureView')
+        # Default to using canvas extents if no case below matches
+        analysis_geoextent = viewport_extent
+        if mode_name == 'HazardExposureView':
+            analysis_geoextent = viewport_extent
+
+        elif mode_name == 'HazardExposure':
+            analysis_geoextent = None
+
+        elif mode_name == 'HazardExposureBookmark' or \
```
Comment (ismailsunni) [inline safe/utilities/analysis.py]: Which one is better? Using bracket or backslash for something like this?

My personal taste : don't use backslash, if possible.

**Comment by ismailsunni** [comment]
selection_049
Perhaps we can move the InaSAFE Dock button the left most, so that Analysis Area button and Scenario Outline button is adjacent to each other.

**Comment by ismailsunni** [comment]
If we turn off the Toggle Scenario Outline and we choose `use intersection of hazard, exposure, and this bounding box`, InaSAFE will still run based on `use intersection of hazard, exposure, and this bounding box`.
Is this expected outcome? Because, if the user set the bbox not inside the intersection hazard and exposure, it will raise InsufficientOverlapError. This is not so natural (at least for me).

**Comment by ismailsunni** [comment]
Besides my comments, all is good to me. I have tried the tools, and so far so good.

**Comment by Charlotte-Morgan** [comment]
+ 1 for dock icon to the left.
- [ ] the window does not close when i click ok. i sort of expected that it would
- [ ] with the first radio button - Use intersection of hazard, exposure and current view extent - i expect the potential analysis (exposure & hazard) extent to be clipped to the window extent. If i either zoom in or move to the side I still get only the intersection of hazard and exposure layers (ie it ignores the window extent)
- [ ] if i use the second radio button it behaves as expected
- [ ] if i use the third radio button H + e + bookmark - i am unable to define a new bookmark - should i expect to be able to define one from here?
- [ ] if i use the 4th radio button i do not get the drag on map option

:( - i checked out the pull request but i feel like i'm looking at the wrong thing?

**Comment by timlinux** [comment]
@Charlotte-Morgan 

> if i use the third radio button H + e + bookmark - i am unable to define a new bookmark - should i expect to be able to define one from here?

No the idea is that you use the bookmark tool built into QGIS to define your bookmarks.

**Comment by timlinux** [comment]
@Charlotte-Morgan to try out the latest code do:

```
git remote add timlinux git@github.com:timlinux/inasafe.git
git checkout timlinux-insafe-1811
git fetch timlinux
git reset --hard timlinux/inasafe-1811
```

**Comment by Charlotte-Morgan** [comment]
@timlinux - thanks for the help above. I now have hacked into your system and I can make book marks. :)
The extent selector works fine for all Jakarta flood raster on raster options .
Re the bookmarks. if none are defined the box has greyed out text "new bookmark" which made me think i could define a bookmark from within the extent selector.

**Comment by timlinux** [comment]
@Charlotte-Morgan can you confirm if you still have the above issues after pulling the latest copy of my code?

**Comment by Charlotte-Morgan** [comment]
@timlinux - no issues with the PR - though only tested on limited data.
So everything works fine - User was a little confused by the new bookmark - see last comment.

**Comment by timlinux** [comment]
@Charlotte-Morgan ok thanks - I am going to merge this PR now and will do another tweak to make the bookmarks more obvious after.

**Comment by timlinux** [comment]
@Charlotte-Morgan  In 48a4e0c I am adding a little more logic to disable the bookmarks radio and show a message if there are no bookmarks defined:

screen shot 2015-08-21 at 13 12 08
Above: UI when user does have bookmarks

screen shot 2015-08-21 at 13 11 47
Above: UI when user has no bookmarks

**Pull request merged**